packages/material-ui/src/Badge/Badge.js

```diff
@@ -1,258 +1,203 @@
-    left: '14%',
-    transform: 'scale(1) translate(-50%, 50%)',
-    transformOrigin: '0% 100%',
-    '&$invisible': {
-      transform: 'scale(0) translate(-50%, 50%)',
-    },
-  },
-  /* Pseudo-class to the badge `span` element if `invisible={true}`. */
-  invisible: {
-    transition: theme.transitions.create('transform', {
-      easing: theme.transitions.easing.easeInOut,
-      duration: theme.transitions.duration.leavingScreen,
-    }),
-  },
-});
-
-const usePreviousProps = (value) => {
-  const ref = React.useRef({});
-  React.useEffect(() => {
-    ref.current = value;
-  });
-  return ref.current;
-};
-
-const Badge = React.forwardRef(function Badge(props, ref) {
+const overridesResolver = (props, styles) => {
   const {
-    anchorOrigin: anchorOriginProp = {
+    color = 'default',
+    variant = 'standard',
+    anchorOrigin = {
       vertical: 'top',
       horizontal: 'right',
     },
-    badgeContent: badgeContentProp,
-    children,
-    classes,
-    className,
-    color: colorProp = 'default',
-    component: ComponentProp = 'span',
-    invisible: invisibleProp,
-    max: maxProp = 99,
-    overlap: overlapProp = 'rectangular',
-    showZero = false,
-    variant: variantProp = 'standard',
-    ...other
+    invisible,
+    overlap = 'rectangular',
   } = props;
 
-  const prevProps = usePreviousProps({
-    anchorOrigin: anchorOriginProp,
-    badgeContent: badgeContentProp,
-    color: colorProp,
-    max: maxProp,
-    overlap: overlapProp,
-    variant: variantProp,
-  });
-
-  let invisible = invisibleProp;
-
-  if (
-    invisibleProp == null &&
-    ((badgeContentProp === 0 && !showZero) || (badgeContentProp == null && variantProp !== 'dot'))
-  ) {
-    invisible = true;
-  }
+  const styleOverrides = {
+    ...styles.root,
+    [`& .${badgeClasses.badge}`]: {
+      ...styles.badge,
+      ...styles[variant],
+      ...styles[
+        `anchorOrigin${capitalize(anchorOrigin.vertical)}${capitalize(
+          anchorOrigin.horizontal,
+        )}${capitalize(overlap)}`
+      ],
+      ...(color !== 'default' && styles[`color${capitalize(color)}`]),
+      ...(invisible && styles.invisible),
+    },
+  };
 
-  const {
-    anchorOrigin = anchorOriginProp,
-    badgeContent,
-    color = colorProp,
-    max = maxProp,
-    overlap = overlapProp,
-    variant = variantProp,
-  } = invisible ? prevProps : props;
+  return styleOverrides;
+};
 
-  const themeVariantsClasses = useThemeVariants(
-    {
-      ...props,
-      anchorOrigin,
-      badgeContent,
-      color,
-      component: ComponentProp,
-      invisible,
-      max,
-      overlap,
-      showZero,
-      variant,
-    },
-    'MuiBadge',
-  );
+const BadgeRoot = styled(
+  'span',
+  {},
+  { muiName: 'MuiBadge', overridesResolver },
+)({
+  position: 'relative',
+  display: 'inline-flex',
+  // For correct alignment with the text.
+  verticalAlign: 'middle',
+  flexShrink: 0,
+});
 
-  let displayValue = '';
+const BadgeBadge = styled(
```

Comment:
This looks really ugly :( I wonder if we should rename the `BadgeRoot` to `BadgeWrapper` and the `BadgeBadge` to `BadgeRoot`

Reply:
With this we currently have a bug, where the theme's variants go to the Root instead of the badge component, which is wrong with this component..

Reply:
From my perspective, the current names are as good as they could get. I think that it surfaces that we could change the composition API of the component. The current approach is a bit weird.

Reply:
We could also leave this for a follow-up PR so that we don't pollute this one

Reply:
I will leave this comment open and we can resolve it after this one is merged, so we will avoid unnecessary changes as part of this PR. Would that work?

Reply:
💯, the point around the weird API isn't specific to unstyled. It's more about how `<Badge>` could be a `<BadgeContainer>`, it's not really a badge as developers can find in other UI libraries.